### PR TITLE
Make rng an argument to ds

### PR DESCRIPTION
**Pull request by hovatterz**
I wanted to pass in a seeded rng. I'm not sure how to test the wasm build, but it's working great as a Rust crate. Hopefully this helps someone out in the future!

### CODE REVIEW

**Comment by fralonra** [comment]
Hi, thanks for your contribution.

It's a great idea to pass rng as an argument.
But sometimes people might not care about the rng, passing in the rng everytime would be verbose.
In my opinion, it would be better to allow user to change the rng rather than forcing them to pass in rng.

For example:
```rust
let mut runner = Runner::with_rng(rng);

// or
let mut runner = Runner::new();
runner.set_rng(rng);
```

**Comment by hovatterz** [comment]
I tried a few different angles to implement that, but I think that might be beyond me or it's too early in the morning ☕. Couldn't find a good way to store a mutable reference to `dyn rand::RngCore` while also allowing a default of `rand::thread_rng()` without making a mess of things.

Any thoughts?

**Comment by fralonra** [comment]
How about these:
```rust
/// Run the Diamond-square algorithm.
pub fn ds(&mut self) -> Output; // Use a default rng

/// Run the Diamond-square algorithm with given rng.
pub fn ds_with_rng(&mut self, rng: &mut impl Rng) -> Output;
```

Then the rng should be passed to downstream, and no need to store it anymore.

File: rs/src/lib.rs [hunk 1]
```diff
@@ -122,23 +120,26 @@ impl Runner {
...
             side: 0,
-            rng: rand::thread_rng(),
         };
         runner.set_width(DEFAULT_WIDTH);
         runner.set_height(DEFAULT_HEIGHT);
         runner
     }
-
-    /// Run the Diamond-square algorithm.
+    /// Run the Diamond-square algorithm with a default rng.
```
Comment (fralonra) [inline rs/src/lib.rs:128]: Insert an empty line above.

File: rs/src/lib.rs [hunk 2]
```diff
@@ -122,23 +120,26 @@ impl Runner {
...
     pub fn ds(&mut self) -> Output {
+        self.ds_with_rng(&mut rand::thread_rng())
+    }
+
+    /// Run the Diamond-square algorithm with given rng.
+    pub fn ds_with_rng(&mut self, rng: &mut impl rand::Rng) -> Output {
```
Comment (fralonra) [inline rs/src/lib.rs:134]: `rand::Rng` is already imported, so just `Rng` here.

File: rs/src/lib.rs [hunk 3]
```diff
@@ -197,7 +198,7 @@ impl Runner {
         }
     }
 
-    fn diamond(&mut self, x: f32, y: f32, half_w: f32, half_h: f32) {
+    fn diamond(&mut self, x: f32, y: f32, half_w: f32, half_h: f32, rng: &mut impl rand::Rng) {
```
Comment (fralonra) [inline rs/src/lib.rs:201]: `rand::Rng` is already imported, so just `Rng` here.

File: rs/src/lib.rs [hunk 4]
```diff
@@ -235,8 +236,8 @@ impl Runner {
         (x == 0.0 && y == 0.0) || (x == 0.0 && y == p) || (x == p && y == 0.0) || (x == p && y == p)
     }
 
-    fn randomize(&mut self, base: f32, range: f32) -> f32 {
-        let r: f32 = self.rng.gen();
+    fn randomize(&mut self, base: f32, range: f32, rng: &mut impl rand::Rng) -> f32 {
```
Comment (fralonra) [inline rs/src/lib.rs:239]: `rand::Rng` is already imported, so just `Rng` here.

File: rs/src/lib.rs [hunk 5]
```diff
@@ -284,22 +285,22 @@ impl Runner {
...
             }
             y += half_h;
         }
 
-        self.shape((size_w / 2.0).ceil(), (size_h / 2.0).ceil());
+        self.shape((size_w / 2.0).ceil(), (size_h / 2.0).ceil(), rng);
     }
 
-    fn square(&mut self, x: f32, y: f32, half_w: f32, half_h: f32) {
+    fn square(&mut self, x: f32, y: f32, half_w: f32, half_h: f32, rng: &mut impl Rng) {
```
Comment (fralonra) [inline rs/src/lib.rs:297]: `rand::Rng` is already imported, so just `Rng` here.

**Review by fralonra — CHANGES_REQUESTED** [review]
Unify coding style.

**Comment by fralonra** [comment]
> I sure overthought that one eh? Should be good to go now.

Sorry for the strictness, but I would like you to make more changes about the coding style.

Btw, it would be better if you could update the document for this feature! Thanks.

**Comment by hovatterz** [comment]
No problem! I've been messing with this while working on my project, so I missed a few things there. Changes have been made. This doesn't apply to the WASM version so I updated the README in the rs dir.

File: rs/README.md [hunk 6]
```diff
@@ -27,4 +27,22 @@ fn main() {
...
+```
+
+If you would like to supply a random number generator:
+
+```rust
+use ds_heightmap::Runner;
+use rand_chacha;
+
+fn main() {
+    let mut rng = ChaCha8Rng::from_entropy();
```
Comment (fralonra) [inline rs/README.md:39]: 1. `ChaCha8Rng` is undeclared here.
2. `from_entropy` not found in `ChaCha8Rng`.

This example is not runnable.

**Review by fralonra — CHANGES_REQUESTED** [review]
The example in Readme is not runnable

**Comment by fralonra** [comment]
> No problem! I've been messing with this while working on my project, so I missed a few things there. Changes have been made. This doesn't apply to the WASM version so I updated the README in the rs dir.

The exmaple you provided is not runnable.
Could you plz update it so that it can be copied and run without error?

**Comment by fralonra** [comment]
Thanks for your patience.

### One more suggestion:
Use another random number generator instead of `ThreadRng`.

### Why
Since the `Runner::ds` is using `ThreadRng` internally, if we supply it into `Runner::ds_with_rng` again, users might not be awared of the fact that `Runner::ds()` is already using `ThreadRng`. Maybe some notes should be added. I can handle this, dont't worry.

### How
Previous example using `rand_chacha` is nice, except it's not runnable.
It can't compiled because you forgot to import something.
FYR, the below code compiles successfully.
```rust
use ds_heightmap::Runner;
use rand_chacha::{rand_core::SeedableRng, ChaCha8Rng};

fn main() {
    let mut rng = ChaCha8Rng::from_entropy();

    let mut runner = Runner::new();
    let output = runner.ds_with_rng(&mut rng);

    println!("data: {:?}", output.data);
    println!("max: {}", output.max);
    println!("min: {}", output.min);
}

```

**Comment by hovatterz** [comment]
No worries! I'm a bit distracted while trying to make these revisions so my bad on that.

**Comment by fralonra** [comment]
Great works!
Thanks you for your contribution!

**Review by fralonra — COMMENTED** [review]
Merged

**Pull request merged**